lib/charms/grafana_agent/v0/cos_agent.py

```diff
@@ -302,6 +327,8 @@ def __init__(
             charm: The `CharmBase` instance that is instantiating this object.
             relation_name: The name of the relation to communicate over.
             metrics_endpoints: List of endpoints in the form [{"path": path, "port": port}, ...].
+                A callable that returns the list may be passed in case the endpoints need to be
+                generated dynamically.
```

Comment:
- Would you be able to utilize the `refresh_events` arg instead of adding the `Callable` functionality?
- The type hint for the constructor arg needs to be update accordingly. I'm on the fence for whether we should bump LIBAPI and only use a callable instead of a Union.

Reply:
* I don't think so. Not reasonably at least. If you are doing significant logic, you should be doing it in a hook and not in the charm's __init__() method.
* You are suggesting requiring it to be callable? I think most charms will just have a static list so we should allow that.

Reply:
Even in the case where a refresh event would be useful, it still seems to reverse the logic. I can imagine something like this:

```py
def __init__(self):
  self.on($some_relation_change, self.update_cached_thing)
  self._cos = COSAgentProvider(self, refresh_events=[self.on.$some_relation_change], metrics_endpoints=[{some data with self._state.cached_thing}])

def update_cached_thing(self, event):
  self._state.cached_thing = event.data[event.app]["something"]
```

This is a lot harder to grasp at first sight. Further, even like this, the `metrics_endpoints` have already been evaluated, therefore the changes to `self._state.cached_thing` are ignored.

I'd much rather prefer the callable approach, clearly demonstrating the intent of the code, while also making it much easier to test and reason about.

What did you have in mind @sed-i?

Reply:
This is also similar to the `lookaside_jobs` in the `MetricsEndpointProvider` library

Reply:
Yep. Let's keep the callable.
The other approach is events: the `cache_thing` would be expected to emit an event that would be passed as a refresh event.
Resolving.